internal/workflow/reconciler.go

```diff
@@ -2,171 +2,72 @@ package workflow
 
 import (
 	"context"
-	"fmt"
 	"time"
 
-	"github.com/go-logr/logr"
-	"github.com/tinkerbell/tink/api/v1alpha1"
-	"k8s.io/apimachinery/pkg/api/equality"
+	tinkv1 "github.com/tinkerbell/tink/api/v1alpha2"
+	"github.com/tinkerbell/tink/internal/workflow/internal"
 	"k8s.io/apimachinery/pkg/api/errors"
-	"knative.dev/pkg/ptr"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
-	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-// Reconciler is a type for managing Workflows.
+// Reconciler reconciles Workflow instances.
 type Reconciler struct {
-	client  ctrlclient.Client
+	client  client.Client
 	nowFunc func() time.Time
 }
 
-func NewReconciler(client ctrlclient.Client) *Reconciler {
+// NewReconciler creates a Reconciler instance.
+func NewReconciler(clnt client.Client) *Reconciler {
 	return &Reconciler{
-		client:  client,
+		client:  clnt,
 		nowFunc: time.Now,
 	}
 }
 
 // +kubebuilder:rbac:groups=tinkerbell.org,resources=hardware;hardware/status,verbs=get;list;watch;update;patch
 // +kubebuilder:rbac:groups=tinkerbell.org,resources=templates;templates/status,verbs=get;list;watch;update;patch
-// +kubebuilder:rbac:groups=tinkerbell.org,resources=workflows;workflows/status,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=tinkerbell.org,resources=workflows;workflows/status,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=tinkerbell.org,resources=workflows;workflows/finalizers,verbs=update
 
-func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (result reconcile.Result, rerr error) {
```

Comment:
are we gaining any value by having named return values? I don't see them being used, feels unnecessary.

Reply:
The named returns are included for the defer to work because we change the return variable if patching fails. I don't think the code needs to work this way, and honestly I'm not fond of the defer approach - defering the patch is a common method for always patching.

Reply:
Well, that's a good pont: `_`.